Update Readme (#1838)
microsoft/ApplicationInsights-dotnet: .docs/concepts.md

# Application Insights DotNet SDK Concepts

> This lists the high level concepts of our SDK and links to detailed guides to help you get started.

Reviewer: Minor: consider change `our` to a more specific word.

To use the Application Insights SDK you must configure an Instrumentation Key which can be [obtained from an Application Insights resource](https://docs.microsoft.com/azure/azure-monitor/app/create-new-resource).
Or you can use a [Connection string](https://docs.microsoft.com/azure/azure-monitor/app/sdk-connection-string?tabs=net) to simplify the configuration of our endpoints.
Both an Instrumentation Key and Connection String are provided for you on the Overivew Dashboard of your Application Insights resource.

## TelemetryClient
The `TelemetryClient` object is the primary root object for the library. 
Almost all functionality around telemetry sending is located on this object. 

### Initialization
You must initialize an instance of this object and populate it with your Instrumentation Key to identify your data.

```C#
using Microsoft.ApplicationInsights;
using Microsoft.ApplicationInsights.Extensibility;

var configuration = new TelemetryConfiguration
{

>     // Please use either an Instrumentation Key or a Connection String but not both.

Reviewer: Given the direction is to use Connection String, I think we should remove Instrumentation Key from the conceptual document.

Author: I'm trying to ignore my bias here because I implemented Connection Strings. :)

The UX still displays the Instrumentation Key on the Overview blade. Since it's a very simple way to get started with our SDK I think it's still a valid example for getting started. 
I don't feel strongly about it, but most of our official docs still use the ikey example for getting started. 

Maybe this should be a larger refactoring effort for the docs team to review all public docs and official repos?

Reviewer: We can start here. Need help from @ank3it to drive other docs to catch up.

The general principle is that we should only give *one recommended path* in `conceptual documents`. It is okay to list _all possible ways_ in `reference document`.

Author: Thanks for clarifying conceptual vs reference. I'll keep that in mind going forward!

    InstrumentationKey = "YOUR INSTRUMENTAATION KEY",
    ConnectionString = "YOUR CONNECTION STRING",
};
var tc = new TelemetryClient(configuration);
```

### Using the TelemetryClient to send telemetry
You can populate common context on the `TelemetryClient.context` property which will be automatically attached to each telemetry item sent. 
You can also attach additional property data to each telemetry item sent. 
The `TelemetryClient` also exposes a number of `Track` methods that can be used to send all telemetry types understood by the Application Insights service. Some example use cases are shown below.

```C#

> tc.Context.User.Id = Environment.GetUserName(); // This is probably a bad idea from a PII perspective.

Reviewer: Probably replace this with a proper example.

tc.Context.Device.OperatingSystem = Environment.OSVersion.ToString();

tc.TrackPageView("Form1");

> tc.TrackEvent("PurchaseOrderSubmitted", new Dictionary<string, string>() { {"CouponCode", "JULY2015" } }, new Dictionary<string, double>() { {"OrderTotal", 68.99 }, {"ItemsOrdered", 5} });

Reviewer: Change this to a better formatted code snippet instead of single line.

	
try
{

> 	...

Reviewer: Use 4 spaces instead of a tab.

Author: How did you even see that? o_O

}
catch(Exception e)
{
	tc.TrackException(e);
}
``` 

### TelemetryClient API

> The TelemetryClient provides several Track methods to collect different types of logs

Reviewer: ```suggestion
The `TelemetryClient` provides several Track methods to collect different types of telemetry data.
```

> Please review our [Api summary for custom events and metrics](https://docs.microsoft.com/azure/azure-monitor/app/api-custom-events-metrics).

Reviewer: ```suggestion
Please review the [API summary for custom events and metrics](https://docs.microsoft.com/azure/azure-monitor/app/api-custom-events-metrics).
```

> 

Reviewer: Nit: remove extra empty lines here.

## Telemetry Channels

> Telemetry channels are responsible for buffering telemetry items and sending them to the Application Insights service, where they're stored for querying and analysis.

Reviewer: Not necessary to have `buffering`?

Author: I didn't write any of these statements. That's an exact quote from our public doc. Do you think it should say something else here?

Reviewer 2: We can kill buffering in the public docs as well if needed. A lot of the core terminology used in the public docs predates all of us working on the product so if there is something that doesn't make sense, definitely let me know so we can clean it up.

Reviewer: I wonder if this would help to clarify "Telemetry channels are responsible for sending the telemetry data to the designated place. Optional features can be provided by the telemetry channels, for example, buffering the data and sending in them in batches, persisting the data to a local storage in case of transmission failure (e.g. network outage), traffic shaping and retry mechanisms".

Author: How about:
"Telemetry channels are responsible for sending telemetry data. Optional features include buffering and batching data, local data storage in case of transmission failure, and retry mechanisms."

Reviewer 3: Why dont we just link to public doc here? for every topic where there is a public doc, i'd suggest to simply link it here and not write anything more.

Author: @cijothomas, i've given every section an intro statement, copied from the public doc and also a link to that public doc. 
The purpose is if someone knew nothing about the sdk, they could navigate high-level concepts all in one place.
Similar to this: https://github.com/microsoft/ApplicationInsights-Home/issues/256

> The .NET and .NET Core versions of the SDKs provide two built-in telemetry channels::

Reviewer: ```suggestion
The .NET and .NET Core versions of the SDKs provide two built-in telemetry channels:
```
